### PR TITLE
Package Preprocessor6

### PR DESCRIPTION
**Pull request by knewbury01**
## Description

add DIR-4-9

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [x] Queries have been added for the following rules:
     - DIR-4-9
 - [ ] Queries have been modified for the following rules:
     - _rule number here_

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [ ] Yes
  - [x] No

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [ ] Have all the relevant rule package description files been checked in?
 - [ ] Have you verified that the metadata properties of each new query is set appropriately?
 - [ ] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [ ] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [ ] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [ ] Does the query have an appropriate level of in-query comments/documentation?
 - [ ] Have you considered/identified possible edge cases?
 - [ ] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### CODE REVIEW

**Comment by knewbury01** [comment]
# Notes:

this query has a lot of results, I think if there are more omission edge cases that would help, or additional/better heuristics for how to identify good cases to suggest where the rule is applicable. As it stands, I have added it as audit since it has so many results

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import codingstandards.cpp.FunctionLikeMacro
+import codingstandards.cpp.Naming
+
+predicate isOperator(string possible) {
+  possible in [
+      "+", "-", "*", "/", "%", "^", "&", "|", "~", "!", "=", "<", ">", "+=", "-=", "*=", "/=", "%=",
+      "^=", "&=", "|=", "<<", ">>", ">>=", "<<=", "==", "!=", "<=", ">=", "<=>", "&&", "||", "++",
+      "--", "->*", "->", "()", "[]"
+    ]
+}
```
Comment (rvermeulen) [inline c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:27]: We can simplify this to operators actually used in Macro's with:

```suggestion
possible = any(Operation op).getOperator()
```

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+//cases where we trust the choice
+predicate omission(MacroInvocation i) {
+  i.getFile() instanceof HeaderFile or
```
Comment (rvermeulen) [inline c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:31]: Could you help me understand the first omission?

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  //static storage duration can only be initialized with constant
+  not exists(StaticStorageDurationVariable v | i.getExpr() = v.getAnAssignedValue()) and
+  //function call not allowed in a constant expression (where constant expr is parent)
+  not exists(Expr e |
+    e.isConstant() and
+    not i.getExpr() = e and
+    i.getExpr().getParent+() = e
+  ) and
+  forall(string arg | arg = i.getUnexpandedArgument(_) | exists(Expr e | arg = e.toString()))
+select i, "Macro invocation used when function call would be preferred."
```
Comment (rvermeulen) [inline c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:73]: By using the macro invocation location we will get multiple results for the same macro that can be replaced. We can limit it to the macro location to reduce that.

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql (hunk 1 above)
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:27]: agree, done

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql (hunk 2 above)
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:31]: its a pretty arbitrary heuristic , the idea behind it is to narrow to code (source files) with more churn I guess (assuming thats src > headers)

Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:25]: I have omitted that heuristic now, but I still think there may be too many results as is

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql (hunk 3 above)
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:69]: suggested change applied

**Comment by knewbury01** [comment]
# Notes Oct 18

@rvermeulen thanks for the review,

I have left the `UnsafeMacro` base set in the implementation (~500) results on openpilot as opposed to without (~900 results) but we can rm it if it seems not as needed.

otherwise have refactored greatly! please take another look when you have time :)

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql (hunk 3 above)
Comment (rvermeulen) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql]: Yes indeed, I think this is in the spirit of the directive. Otherwise, we would be encouraging code duplication which doesn't seem like a good suggestion.

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+abstract class IrreplaceableFunctionLikeMacro extends FunctionLikeMacro { }
+
+private class AsmArgumentInvoked extends IrreplaceableFunctionLikeMacro {
+  AsmArgumentInvoked() {
+    any(AsmStmt s).getLocation().subsumes(this.getAnInvocation().getLocation())
+  }
+}
+
+private class OnlyConstantNumericInvoked extends IrreplaceableFunctionLikeMacro {
```
[the remark this answers is not in the document]
Comment (rvermeulen) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:31]: We might want to consider a more complete constant invocation modelling that considers other constant literals.

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    e.isConstant() and
+    not i.getExpr() = e and
+    i.getExpr().getParent+() = e
+  )
+}
+
+from FunctionLikeMacro m
+where
+  not isExcluded(m, Preprocessor6Package::functionOverFunctionLikeMacroQuery()) and
+  not omission(m) and
```
Comment (rvermeulen) [inline c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:97]: If `omission` is replaced with a class.
```suggestion

```

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+abstract class UnsafeMacro extends FunctionLikeMacro { }
+
+class ParameterNotUsedMacro extends UnsafeMacro {
+  ParameterNotUsedMacro() {
+    //parameter not used - has false positives on args that are not used but are substrings of other args
+    exists(string p |
+      p = this.getAParameter() and
+      not this.getBody().regexpMatch(".*(\\s*|\\(|\\)|\\##)" + p + "(\\s*||\\)|\\(|\\##).*")
```
Comment (rvermeulen) [inline c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:67]:
```suggestion
      not this.getBody().regexpMatch(".*(\\s*|\\(|\\)|\\##)" + p + "(\\s*|\\)|\\(|\\##).*")
```

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+private class FunctionLikeMacroWithOperatorArgument extends IrreplaceableFunctionLikeMacro {
+  FunctionLikeMacroWithOperatorArgument() {
+    exists(MacroInvocation mi | mi.getMacro() = this |
+      mi.getUnexpandedArgument(_) = any(Operation op).getOperator()
+    )
+  }
+}
+
+abstract class UnsafeMacro extends FunctionLikeMacro { }
```
Comment (rvermeulen) [inline c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:60]: We should be clear about what constitutes an unsafe macro.

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    not i.getExpr() = e and
+    i.getExpr().getParent+() = e
+  )
+}
+
+from FunctionLikeMacro m
+where
+  not isExcluded(m, Preprocessor6Package::functionOverFunctionLikeMacroQuery()) and
+  not omission(m) and
+  m instanceof UnsafeMacro and
```
Comment (rvermeulen) [inline c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:98]: I believe we will miss legitimate cases by limiting ourselves to the unsafe macros. Since this is an audit query I think we can accept the additional results.

**Review by rvermeulen — CHANGES_REQUESTED** [review]

**Comment by rvermeulen** [comment]
Another case we can't replace is if the macro uses a stringizing or concatenation operator.

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+import cpp
+import codingstandards.c.misra
+import codingstandards.cpp.FunctionLikeMacro
+import codingstandards.cpp.Naming
+
+predicate omission(Macro i) { Naming::Cpp14::hasStandardLibraryMacroName(i.getName()) }
+
+abstract class IrreplaceableFunctionLikeMacro extends FunctionLikeMacro { }
+
```
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:24]: agree, done

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+private class AsmArgumentInvoked extends IrreplaceableFunctionLikeMacro {
```
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:25]: agree, done

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+      mi.getUnexpandedArgument(_).regexpMatch("\\d+")
+    )
+  }
+}
+
+private class KnownIrreplaceableFunctionLikeMacro extends IrreplaceableFunctionLikeMacro {
+  KnownIrreplaceableFunctionLikeMacro() {
+    this.getName() in ["UNUSED", "__has_builtin", "MIN", "MAX"]
+  }
+}
```
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:43]: ah, my bad, rm'd!

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+private class UsedToStaticInitialize extends IrreplaceableFunctionLikeMacro {
```
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:45]: agree, done

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql (hunk 5 above)
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:97]: agree, done, omitted the omission :)

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql (hunk 6 above)
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:67]: good eye! darn copy pasting!

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql (hunk 8 above)
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:98]: ah! after the other improvements there actually arent that many more anyways for starting from just regular function like macros, so agreed!

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql (hunk 7 above)
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:60]: not relevant any longer

File: c/misra/test/rules/DIR-4-9/test.c [hunk 13]
```diff
@@ -0,0 +1,30 @@
+#define MACRO(OP, L, R) ((L)OP(R))   // COMPLIANT
```
Comment (rvermeulen) [inline c/misra/test/rules/DIR-4-9/test.c:1]: Overall, I think we should add more non-compliant cases.

File: c/misra/test/rules/DIR-4-9/test.c [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+#define MACRO2(L, R) (L + R)         // COMPLIANT
+#define MACRO3(L, R) (L " " R " " L) // COMPLIANT
+#define MACRO4(L)                                                              \
+  (L" "                                                                        \
```
Comment (rvermeulen) [inline c/misra/test/rules/DIR-4-9/test.c:5]: The `L` prefix is a bit unfortunate because it creates a wide character string instead of using the `L` parameter. So this test case is not doing what we hope it is doing.

I noticed this because the results of `MACRO2(1,1)` should have constant resulting in an expression when we call `getAnExpandedArgument`, however the result was empty indicating that the macro parameter isn't used.

File: c/misra/test/rules/DIR-4-9/test.c [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+  char *i5 = MACRO4("prefix");
+
+  char *i6 = MACRO4(MACRO2(1, 1));
+
+  int i7 = MACRO5(1, 1);
+
+  int i8 = MACRO6(1, 1);
+
+  char *i9 = MACRO7("prefix", "suffix");
```
Comment (rvermeulen) [inline c/misra/test/rules/DIR-4-9/test.c:29]: This test case is causing the compiler error: `error: pasting formed '"prefix""suffix"', an invalid preprocessing token`

**Review by rvermeulen — CHANGES_REQUESTED** [review]

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+import cpp
+import codingstandards.c.misra
+import codingstandards.cpp.FunctionLikeMacro
+import codingstandards.cpp.Naming
+
+abstract class IrreplaceableFunctionLikeMacro extends FunctionLikeMacro { }
+
+/** A standard library function like macro that contains the use of a stringize or tokenize operator should not be replaced by a function. */
+private class StringizeOrTokenizeMacro extends IrreplaceableFunctionLikeMacro {
+  StringizeOrTokenizeMacro() { this.getBody().regexpMatch(".*\\#{1,2}?.*") }
```
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:25]: aha! I thought that those were around, but forgot where, thanks! applied

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+/** A function like macro invocation as an `asm` argument cannot be replaced by a function. */
+private class AsmArgumentInvoked extends IrreplaceableFunctionLikeMacro {
+  AsmArgumentInvoked() {
+    any(AsmStmt s).getLocation().subsumes(this.getAnInvocation().getLocation())
+  }
+}
+
+/** A macro that is only invoked with constant arguments is more likely to be compile-time evaluated than a function call so do not suggest replacement. */
+private class OnlyConstantNumericInvoked extends IrreplaceableFunctionLikeMacro {
+  OnlyConstantNumericInvoked() {
```
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:42]: agree, done

File: c/misra/test/rules/DIR-4-9/test.c [hunk 18]
```diff
@@ -0,0 +1,30 @@
+#define MACRO(OP, L, R) ((L)OP(R))   // COMPLIANT
+#define MACRO2(L, R) (L + R)         // COMPLIANT
+#define MACRO3(L, R) (L " " R " " L) // COMPLIANT
+#define MACRO4(L)                                                              \
+  (L" "                                                                        \
+   "suffix")                   // NON_COMPLIANT
+#define MACRO5(L, LR) (LR + 1) // COMPLIANT
+#define MACRO6(X, LR) (LR + 1) // COMPLIANT
+#define MACRO7(x, y) x##y      // COMPLIANT
```
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/test/rules/DIR-4-9/test.c:9]: agree, done

File: c/misra/test/rules/DIR-4-9/test.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/test/rules/DIR-4-9/test.c:29]: agree, done

Comment (knewbury01) [reply c/misra/test/rules/DIR-4-9/test.c:29]: good catch, fixed

File: c/misra/test/rules/DIR-4-9/test.c (hunk 14 above)
Comment (knewbury01) [reply c/misra/test/rules/DIR-4-9/test.c:5]: ah, another good catch! fixed

File: c/misra/test/rules/DIR-4-9/test.c (hunk 13 above)
Comment (knewbury01) [reply c/misra/test/rules/DIR-4-9/test.c:1]: sure, I added one other case, and actually one more compliant (when there is no body, I think that wont make sense to replace).

I can't think of any other edge cases to represent though, so if you had any suggestions ?

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+ */
+
+import cpp
+import codingstandards.c.misra
+import codingstandards.cpp.Naming
+import codingstandards.cpp.Macro
+
+abstract class IrreplaceableFunctionLikeMacro extends FunctionLikeMacro { }
+
+/** A standard library function like macro that contains the use of a stringize or tokenize operator should not be replaced by a function. */
```
[the remark this answers is not in the document]
Comment (rvermeulen) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:23]: I forgot to mention we should place the `IrreplaceableFunctionLikeMacro` in a library module so we can customize it using Misra's `Customizations.qll`

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+
+from FunctionLikeMacro m
+where
+  not isExcluded(m, Preprocessor6Package::functionOverFunctionLikeMacroQuery()) and
+  not m instanceof IrreplaceableFunctionLikeMacro and
+  //macros can have empty body
+  not m.getBody().length() = 0 and
+  //function call not allowed in a constant expression (where constant expr is parent)
+  forall(MacroInvocation i | i = m.getAnInvocation() | not partOfConstantExpr(i))
+select m, "Macro used when function call would be preferred.", m.getBody().length()
```
Comment (rvermeulen) [inline c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:89]: Consider a rewrite of the alert message to avoid a value judgement.
The current alert message has a spurious argument `m.getBody().length()`

```suggestion
select m, "Macro used instead of a function."
```

File: c/misra/test/rules/DIR-4-9/test.c [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+#define MACRO5(L, LR) (LR + 1)                          // COMPLIANT
+#define MACRO6(x) printf_custom("output = %d", test##x) // COMPLIANT
+#define MACRO7(x) #x                                    // COMPLIANT
+#define MACRO8(x) "NOP"                                 // COMPLIANT
+#define MACRO9() printf_custom("output = %d", 7)        // NON_COMPLIANT
+#define MACRO10(x)                                      // COMPLIANT
+
+const char a1[MACRO2(1, 1) + 6];
+extern printf_custom();
+int test1;
```
Comment (rvermeulen) [inline c/misra/test/rules/DIR-4-9/test.c:14]: Spurious definition.
```suggestion
```

**Review by rvermeulen — CHANGES_REQUESTED** [review]

File: c/misra/test/rules/DIR-4-9/test.c (hunk 13 above)
Comment (rvermeulen) [reply c/misra/test/rules/DIR-4-9/test.c]: Maybe some FNs like `#define MY_ASSERT(X) assert(x)` to show what we don't handle currently?

File: c/misra/test/rules/DIR-4-9/test.c (hunk 21 above)
Comment (knewbury01) [reply c/misra/test/rules/DIR-4-9/test.c:14]: oh, its actually needed for the expansion of MACRO6, without it we get the compiler error:
`line 30: error: identifier "test1" is undefined`

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql (hunk 20 above)
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:89]: good catch, done! :)

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql (hunk 19 above)
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql:23]: sure thing! I think its pretty specific to this rule so I put it in `c/misra/src/rules/DIR-4-9/`, but lmk if you think it should be in an upper common dir

File: c/misra/test/rules/DIR-4-9/test.c (hunk 13 above)
Comment (knewbury01) [reply c/misra/test/rules/DIR-4-9/test.c]: added that one! good point, that is a case where it should be just the function call.

Comment (rvermeulen) [reply c/misra/test/rules/DIR-4-9/test.c]: Was a completely reasonable ask because my comment should have provided more information 😅.

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql (hunk 19 above)
[the remark this answers is not in the document]
Comment (rvermeulen) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql]: We have to, unfortunately, move it to the common pack because QL doesn't support importing modules with a `-` in the path 😭.

File: c/misra/test/rules/DIR-4-9/test.c (hunk 21 above)
Comment (rvermeulen) [reply c/misra/test/rules/DIR-4-9/test.c:17]: Ah, completely right. Will compile check next time.

File: c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql (hunk 19 above)
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql]: sure, thing moved to `c/common/src/codingstandards/c`

**Review by rvermeulen — APPROVED** [review]
LGTM
